internal/locate/region_cache.go

```diff
@@ -508,6 +508,132 @@ func (mu *regionIndexMu) refresh(r []*Region) {
 
 type livenessFunc func(ctx context.Context, s *Store) livenessState
 
+// repeat wraps a `func()` as a schedulable fuction for `bgRunner`.
+func repeat(f func()) func(context.Context, time.Time) bool {
+	return func(_ context.Context, _ time.Time) bool {
+		f()
+		return false
+	}
+}
+
+// until wraps a `func() bool` as a schedulable fuction for `bgRunner`.
+func until(f func() bool) func(context.Context, time.Time) bool {
+	return func(_ context.Context, _ time.Time) bool {
+		return f()
+	}
+}
+
+type bgRunner struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func newBackgroundRunner(ctx context.Context) *bgRunner {
+	ctx, cancel := context.WithCancel(ctx)
+	return &bgRunner{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+}
+
+func (r *bgRunner) closed() bool {
+	select {
+	case <-r.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (r *bgRunner) shutdown(wait bool) {
+	r.cancel()
+	if wait {
+		r.wg.Wait()
+	}
+}
+
+// run calls `f` once in background.
+func (r *bgRunner) run(f func(context.Context)) {
+	if r.closed() {
+		return
+	}
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		f(r.ctx)
+	}()
+}
+
+// schedule calls `f` every `interval`.
+func (r *bgRunner) schedule(f func(context.Context, time.Time) bool, interval time.Duration) {
+	if r.closed() {
+		return
+	}
+	ticker := time.NewTicker(interval)
+	r.wg.Add(1)
+	go func() {
+		defer func() {
+			r.wg.Done()
+			ticker.Stop()
+		}()
+		for {
+			select {
+			case <-r.ctx.Done():
+				return
+			case t := <-ticker.C:
+				if f(r.ctx, t) {
+					return
+				}
+			}
+		}
+	}()
+}
+
+// scheduleWithTrigger likes schedule, but also call `f` when `<-trigger`, in which case the time arg of `f` is zero.
+func (r *bgRunner) scheduleWithTrigger(f func(context.Context, time.Time) bool, interval time.Duration, trigger <-chan struct{}) {
+	if r.closed() {
+		return
+	}
+	ticker := time.NewTicker(interval)
+	r.wg.Add(1)
+	go func() {
+		defer func() {
+			r.wg.Done()
+			ticker.Stop()
+		}()
+		triggerEnabled := trigger != nil
+		for triggerEnabled {
```

Comment:
Why use a standalone for loop for the trigger events instead of putting it in the for loop below?

Reply:
The `trigger` is provided by the caller and we can not make sure it won't be closed by accident. If it closed, there will be a busy loop. Once `trigger` get closed, it should be considered asthere is  no more events.

Reply:
So the purpose is to avoild `select` on an already closed channel right?

Reply:
Yes. Actually `notifyCheckCh` never gets closed, just try to make the runner robust here.